File: src/peer.rs
//! Peer management and connection state tracking

use std::collections::{HashMap, HashSet, VecDeque};
use std::net::SocketAddr;
use std::time::{Duration, Instant};
use crate::crypto::NodeIdentity;
use crate::message::{Endpoint, DisconnectReason};
use crate::error::{Error, Result};

/// Maximum age of peer information before eviction
pub const PEER_TTL: Duration = Duration::from_secs(24 * 60 * 60); // 24 hours

/// Maximum peers to share in PEX message
pub const MAX_PEX_PEERS: usize = 50;

/// Connection state for a peer
#[derive(Debug, Clone)]
pub struct Peer {
    /// Node ID of the peer
    pub node_id: [u8; 32],
    /// Remote address
    pub remote_addr: SocketAddr,
    /// Advertised listening address
    pub listen_addr: Option<SocketAddr>,
    /// Supported protocols
    pub protocols: Vec<String>,
    /// Connection state
    pub state: PeerState,
    /// When the connection was established
    pub connected_at: Option<Instant>,
    /// Last activity timestamp
    pub last_activity: Instant,
    /// Round-trip latency measurement
    pub latency: Option<Duration>,
    /// Trust score (0-100)
    pub trust_score: u8,
}

impl Peer {
    pub fn new(node_id: [u8; 32], remote_addr: SocketAddr) -> Self {
        Self {
            node_id,
            remote_addr,
            listen_addr: None,
            protocols: Vec::new(),
            state: PeerState::Connecting,
            connected_at: None,
            last_activity: Instant::now(),
            latency: None,
            trust_score: 50, // Default neutral trust
        }
    }
    
    /// Check if peer is active (connected and responsive)
    pub fn is_active(&self) -> bool {
        matches!(self.state, PeerState::Connected)
    }
    
    /// Check if peer info is stale
    pub fn is_stale(&self) -> bool {
        self.last_activity.elapsed() > PEER_TTL
    }
    
    /// Update last activity timestamp
    pub fn mark_active(&mut self) {
        self.last_activity = Instant::now();
    }
    
    /// Set connection as established
    pub fn set_connected(&mut self, protocols: Vec<String>, listen_addr: Option<SocketAddr>) {
        self.state = PeerState::Connected;
        self.connected_at = Some(Instant::now());
        self.protocols = protocols;
        self.listen_addr = listen_addr;
    }
    
    /// Convert to Endpoint for PEX
    pub fn to_endpoint(&self) -> Option<Endpoint> {
        self.listen_addr.or(Some(self.remote_addr)).map(Endpoint::from_socket_addr)
    }
}

/// Peer connection state
#[derive(Debug, Clone, PartialEq)]
pub enum PeerState {
    /// Initial connection in progress
    Connecting,
    /// Handshake in progress
    Handshaking,
    /// Connection established, ready for messages
    Connected,
    /// Graceful disconnect in progress
    Disconnecting,
    /// Connection closed
    Disconnected,
}

/// Peer manager for tracking all known peers
pub struct PeerManager {
    /// All known peers indexed by node ID
    peers: HashMap<[u8; 32], Peer>,
    /// Peers indexed by remote address for quick lookup
    by_address: HashMap<SocketAddr, [u8; 32]>,
    /// Set of blacklisted node IDs
    blacklist: HashSet<[u8; 32]>,
    /// Set of blacklisted addresses
    blacklist_addr: HashSet<SocketAddr>,
    /// Maximum number of connections
    max_connections: usize,
    /// Our own node identity
    identity: NodeIdentity,
}

impl PeerManager {
    pub fn new(identity: NodeIdentity, max_connections: usize) -> Self {
        Self {
            peers: HashMap::new(),
            by_address: HashMap::new(),
            blacklist: HashSet::new(),
            blacklist_addr: HashSet::new(),
            max_connections,
            identity,
        }
    }
    
    /// Add or update a peer
    pub fn add_peer(&mut self, peer: Peer) {
        let node_id = peer.node_id;
        
        if let Some(existing) = self.peers.get(&node_id) {
            // Update address mapping if changed
            if existing.remote_addr != peer.remote_addr {
                self.by_address.remove(&existing.remote_addr);
            }
        }
        
        self.by_address.insert(peer.remote_addr, node_id);
        self.peers.insert(node_id, peer);
    }
    
    /// Get peer by node ID
    pub fn get_peer(&self, node_id: &[u8; 32]) -> Option<&Peer> {
        self.peers.get(node_id)
    }
    
    /// Get mutable peer by node ID
    pub fn get_peer_mut(&mut self, node_id: &[u8; 32]) -> Option<&mut Peer> {
        self.peers.get_mut(node_id)
    }
    
    /// Get peer by address
    pub fn get_peer_by_addr(&self, addr: &SocketAddr) -> Option<&Peer> {
        self.by_address.get(addr).and_then(|id| self.peers.get(id))
    }
    
    /// Remove peer by node ID
    pub fn remove_peer(&mut self, node_id: &[u8; 32]) -> Option<Peer> {
        if let Some(peer) = self.peers.remove(node_id) {
            self.by_address.remove(&peer.remote_addr);
            Some(peer)
        } else {
            None
        }
    }
    
    /// Mark peer as disconnected
    pub fn disconnect_peer(&mut self, node_id: &[u8; 32]) {
        if let Some(peer) = self.peers.get_mut(node_id) {
            peer.state = PeerState::Disconnected;
            peer.connected_at = None;
        }
    }
    
    /// Add node to blacklist
    pub fn blacklist(&mut self, node_id: &[u8; 32]) {
        self.blacklist.insert(*node_id);
    }
    
    /// Add address to blacklist
    pub fn blacklist_address(&mut self, addr: SocketAddr) {
        self.blacklist_addr.insert(addr);
    }
    
    /// Check if node is blacklisted
    pub fn is_blacklisted(&self, node_id: &[u8; 32]) -> bool {
        self.blacklist.contains(node_id)
    }
    
    /// Check if address is blacklisted
    pub fn is_address_blacklisted(&self, addr: &SocketAddr) -> bool {
        self.blacklist_addr.contains(addr)
    }
    
    /// Get count of active connections
    pub fn active_count(&self) -> usize {
        self.peers.values().filter(|p| p.is_active()).count()
    }
    
    /// Check if we can accept more connections
    pub fn can_accept(&self) -> bool {
        self.active_count() < self.max_connections
    }
    
    /// Get random active peers for PEX (peer exchange)
    pub fn get_random_peers(&self, limit: usize) -> Vec<Endpoint> {
        let active: Vec<&Peer> = self.peers.values()
            .filter(|p| p.is_active())
            .collect();
        
        if active.is_empty() {
            return Vec::new();
        }
        
        use rand::seq::SliceRandom;
        let mut rng = rand::thread_rng();
        let selected: Vec<&Peer> = active.choose_multiple(&mut rng, limit.min(MAX_PEX_PEERS)).collect();
        
        selected.iter()
            .filter_map(|p| p.to_endpoint())
            .collect()
    }
    
    /// Get closest peers to a target node ID (for DHT)
    pub fn get_closest_peers(&self, target: &[u8; 32], k: usize) -> Vec<Peer> {
        let mut peers: Vec<(&Peer, [u8; 32])> = self.peers.values()
            .filter(|p| p.is_active())
            .map(|p| {
                let distance = NodeIdentity::distance(&p.node_id, target);
                (p, distance)
            })
            .collect();
        
        // Sort by distance
        peers.sort_by(|a, b| {
            NodeIdentity::compare_distance(&a.1, &b.1)
        });
        
        peers.into_iter()
            .take(k)
            .map(|(p, _)| p.clone())
            .collect()
    }
    
    /// Get all active peers
    pub fn active_peers(&self) -> Vec<&Peer> {
        self.peers.values().filter(|p| p.is_active()).collect()
    }
    
    /// Evict stale peers
    pub fn evict_stale(&mut self) -> Vec<[u8; 32]> {
        let stale: Vec<[u8; 32]> = self.peers.values()
            .filter(|p| p.is_stale() && !p.is_active())
            .map(|p| p.node_id)
            .collect();
        
        for node_id in &stale {
            self.remove_peer(node_id);
        }
        
        stale
    }
    
    /// Update peer latency measurement
    pub fn update_latency(&mut self, node_id: &[u8; 32], latency: Duration) {
        if let Some(peer) = self.peers.get_mut(node_id) {
            peer.latency = Some(latency);
        }
    }
    
    /// Adjust trust score
    pub fn adjust_trust(&mut self, node_id: &[u8; 32], delta: i8) {
        if let Some(peer) = self.peers.get_mut(node_id) {
            let new_score = (peer.trust_score as i16 + delta as i16).clamp(0, 100) as u8;
            peer.trust_score = new_score;
        }
    }
}

/// Connection event types
#[derive(Debug, Clone)]
pub enum ConnectionEvent {
    /// New inbound connection received
    InboundConnected {
        addr: SocketAddr,
    },
    /// Outbound connection established
    OutboundConnected {
        addr: SocketAddr,
        node_id: [u8; 32],
    },
    /// Connection closed
    Disconnected {
        addr: SocketAddr,
        node_id: Option<[u8; 32]>,
        reason: DisconnectReason,
    },
    /// Handshake completed
    HandshakeComplete {
        node_id: [u8; 32],
        protocols: Vec<String>,
    },
    /// Error during connection
    Error {
        addr: SocketAddr,
        error: Error,
    },
}

/// Pending outbound connections
pub struct PendingConnections {
    /// Connections being established
    pending: HashMap<SocketAddr, Instant>,
    /// Timeout for pending connections
    timeout: Duration,
}

impl PendingConnections {
    pub fn new(timeout: Duration) -> Self {
        Self {
            pending: HashMap::new(),
            timeout,
        }
    }
    
    /// Add a pending connection
    pub fn add(&mut self, addr: SocketAddr) -> bool {
        // Don't add if already pending
        if self.pending.contains_key(&addr) {
            return false;
        }
        
        self.pending.insert(addr, Instant::now());
        true
    }
    
    /// Remove a pending connection
    pub fn remove(&mut self, addr: &SocketAddr) {
        self.pending.remove(addr);
    }
    
    /// Check if connection is pending
    pub fn is_pending(&self, addr: &SocketAddr) -> bool {
        self.pending.contains_key(addr)
    }
    
    /// Clean up timed out pending connections
    pub fn cleanup(&mut self) -> Vec<SocketAddr> {
        let now = Instant::now();
        let timed_out: Vec<SocketAddr> = self.pending.iter()
            .filter(|(_, started)| now.duration_since(**started) > self.timeout)
            .map(|(addr, _)| *addr)
            .collect();
        
        for addr in &timed_out {
            self.pending.remove(addr);
        }
        
        timed_out
    }
}

File: src/peer.cpp
/**
 * Peer Management Implementation - C++
 */

#include "peer.hpp"
#include "crypto.hpp"
#include <algorithm>
#include <random>
#include <sstream>
#include <iomanip>

namespace aether {

// Helper to convert NodeId to hex string
static std::string node_id_to_hex(const NodeId& id) {
    static const char hex_chars[] = "0123456789abcdef";
    std::ostringstream oss;
    for (uint8_t byte : id) {
        oss << hex_chars[(byte >> 4) & 0xF] << hex_chars[byte & 0xF];
    }
    return oss.str();
}

// Peer implementation

Peer::Peer(const NodeId& node_id, const std::string& addr, uint16_t port)
    : node_id(node_id)
    , remote_addr(addr)
    , remote_port(port)
    , last_activity(std::chrono::steady_clock::now())
{}

bool Peer::is_stale() const {
    return std::chrono::steady_clock::now() - last_activity > DEFAULT_TTL;
}

void Peer::mark_active() {
    last_activity = std::chrono::steady_clock::now();
}

void Peer::set_connected(const std::vector<std::string>& protocols,
                         const std::optional<std::string>& listen_addr,
                         const std::optional<uint16_t>& listen_port) {
    state = PeerState::Connected;
    connected_at = std::chrono::steady_clock::now();
    this->protocols = protocols;
    this->listen_addr = listen_addr;
    this->listen_port = listen_port;
}

std::optional<Endpoint> Peer::to_endpoint() const {
    if (listen_addr && listen_port) {
        return Endpoint::from_ipv4(*listen_addr, *listen_port);
    }
    if (!remote_addr.empty() && remote_port > 0) {
        return Endpoint::from_ipv4(remote_addr, remote_port);
    }
    return std::nullopt;
}

// PeerManager implementation

PeerManager::PeerManager(size_t max_connections)
    : max_connections_(max_connections)
{}

void PeerManager::add_peer(const Peer& peer) {
    std::string id_hex = node_id_to_hex(peer.node_id);

    // Remove old address mapping if exists
    auto it = peers_.find(id_hex);
    if (it != peers_.end() && it->second.remote_addr != peer.remote_addr) {
        by_address_.erase(it->second.remote_addr + ":" + std::to_string(it->second.remote_port));
    }

    // Add/update peer
    peers_[id_hex] = peer;

    // Update address mapping
    std::string addr_key = peer.remote_addr + ":" + std::to_string(peer.remote_port);
    by_address_[addr_key] = id_hex;
}

const Peer* PeerManager::get_peer(const NodeId& node_id) const {
    std::string id_hex = node_id_to_hex(node_id);
    auto it = peers_.find(id_hex);
    return (it != peers_.end()) ? &it->second : nullptr;
}

Peer* PeerManager::get_peer(const NodeId& node_id) {
    std::string id_hex = node_id_to_hex(node_id);
    auto it = peers_.find(id_hex);
    return (it != peers_.end()) ? &it->second : nullptr;
}

const Peer* PeerManager::get_peer_by_addr(const std::string& addr, uint16_t port) const {
    std::string addr_key = addr + ":" + std::to_string(port);
    auto it = by_address_.find(addr_key);
    if (it != by_address_.end()) {
        NodeId node_id;
        std::istringstream iss(it->second);
        std::string hex;
        iss >> hex;
        if (hex.length() == NODE_ID_SIZE * 2) {
            for (size_t i = 0; i < NODE_ID_SIZE; i++) {
                node_id[i] = static_cast<uint8_t>(std::stoul(hex.substr(i * 2, 2), nullptr, 16));
            }
            return get_peer(node_id);
        }
    }
    return nullptr;
}

Peer* PeerManager::get_peer_by_addr(const std::string& addr, uint16_t port) {
    std::string addr_key = addr + ":" + std::to_string(port);
    auto it = by_address_.find(addr_key);
    if (it != by_address_.end()) {
        NodeId node_id;
        std::istringstream iss(it->second);
        std::string hex;
        iss >> hex;
        if (hex.length() == NODE_ID_SIZE * 2) {
            for (size_t i = 0; i < NODE_ID_SIZE; i++) {
                node_id[i] = static_cast<uint8_t>(std::stoul(hex.substr(i * 2, 2), nullptr, 16));
            }
            return get_peer(node_id);
        }
    }
    return nullptr;
}

std::optional<Peer> PeerManager::remove_peer(const NodeId& node_id) {
    std::string id_hex = node_id_to_hex(node_id);
    auto it = peers_.find(id_hex);

    if (it != peers_.end()) {
        Peer peer = it->second;
        by_address_.erase(peer.remote_addr + ":" + std::to_string(peer.remote_port));
        peers_.erase(it);
        return peer;
    }

    return std::nullopt;
}

void PeerManager::disconnect_peer(const NodeId& node_id) {
    Peer* peer = get_peer(node_id);
    if (peer) {
        peer->state = PeerState::Disconnected;
        peer->connected_at = std::nullopt;
    }
}

void PeerManager::blacklist(const NodeId& node_id) {
    blacklist_.insert(node_id_to_hex(node_id));
}

void PeerManager::blacklist_address(const std::string& addr) {
    blacklist_addr_.insert(addr);
}

bool PeerManager::is_blacklisted(const NodeId& node_id) const {
    return blacklist_.count(node_id_to_hex(node_id)) > 0;
}

bool PeerManager::is_address_blacklisted(const std::string& addr) const {
    return blacklist_addr_.count(addr) > 0;
}

size_t PeerManager::active_count() const {
    size_t count = 0;
    for (const auto& [id, peer] : peers_) {
        if (peer.is_active()) {
            count++;
        }
    }
    return count;
}

std::vector<Endpoint> PeerManager::get_random_peers(size_t limit) const {
    std::vector<const Peer*> active;
    active.reserve(peers_.size());

    for (const auto& [id, peer] : peers_) {
        if (peer.is_active()) {
            active.push_back(&peer);
        }
    }

    if (active.empty()) {
        return {};
    }

    // Shuffle and take limit
    static std::random_device rd;
    static std::mt19937 gen(rd());

    std::shuffle(active.begin(), active.end(), gen);

    std::vector<Endpoint> result;
    size_t count = std::min(limit, active.size());
    result.reserve(count);

    for (size_t i = 0; i < count; i++) {
        if (auto ep = active[i]->to_endpoint()) {
            result.push_back(*ep);
        }
    }

    return result;
}

std::vector<Peer> PeerManager::get_closest_peers(const NodeId& target, size_t k) const {
    std::vector<std::pair<const Peer*, std::array<uint8_t, NODE_ID_SIZE>>> peers_with_dist;

    for (const auto& [id, peer] : peers_) {
        if (peer.is_active()) {
            auto dist = Crypto::distance(peer.node_id, target);
            peers_with_dist.emplace_back(&peer, dist);
        }
    }

    // Sort by distance
    std::sort(peers_with_dist.begin(), peers_with_dist.end(),
        [](const auto& a, const auto& b) {
            return Crypto::compare_distance(a.second, b.second) < 0;
        });

    // Return top k
    std::vector<Peer> result;
    result.reserve(std::min(k, peers_with_dist.size()));

    for (size_t i = 0; i < k && i < peers_with_dist.size(); i++) {
        result.push_back(*peers_with_dist[i].first);
    }

    return result;
}

std::vector<const Peer*> PeerManager::active_peers() const {
    std::vector<const Peer*> result;
    result.reserve(active_count());

    for (const auto& [id, peer] : peers_) {
        if (peer.is_active()) {
            result.push_back(&peer);
        }
    }

    return result;
}

std::vector<NodeId> PeerManager::evict_stale() {
    std::vector<NodeId> evicted;

    for (auto it = peers_.begin(); it != peers_.end(); ) {
        if (it->second.is_stale() && !it->second.is_active()) {
            evicted.push_back(it->second.node_id);
            by_address_.erase(it->second.remote_addr + ":" + std::to_string(it->second.remote_port));
            it = peers_.erase(it);
        } else {
            ++it;
        }
    }

    return evicted;
}

void PeerManager::update_latency(const NodeId& node_id, Duration latency) {
    Peer* peer = get_peer(node_id);
    if (peer) {
        peer->latency = latency;
    }
}

void PeerManager::adjust_trust(const NodeId& node_id, int delta) {
    Peer* peer = get_peer(node_id);
    if (peer) {
        int new_score = static_cast<int>(peer->trust_score) + delta;
        peer->trust_score = static_cast<uint8_t>(std::clamp(new_score, 0, 100));
    }
}

}  // namespace aether
